pkg/operator/etcdendpointscontroller/etcdendpointscontroller.go

```diff
@@ -0,0 +1,171 @@
+	operatorClient  v1helpers.OperatorClient
+	nodeLister      corev1listers.NodeLister
+	configmapLister corev1listers.ConfigMapLister
+	configmapClient corev1client.ConfigMapsGetter
+}
+
+func NewEtcdEndpointsController(
+	operatorClient v1helpers.OperatorClient,
+	eventRecorder events.Recorder,
+	kubeClient kubernetes.Interface,
+	kubeInformers operatorv1helpers.KubeInformersForNamespaces,
+) factory.Controller {
+	kubeInformersForTargetNamespace := kubeInformers.InformersFor(operatorclient.TargetNamespace)
+	configmapsInformer := kubeInformersForTargetNamespace.Core().V1().ConfigMaps()
+	kubeInformersForCluster := kubeInformers.InformersFor("")
+	nodeInformer := kubeInformersForCluster.Core().V1().Nodes()
+
+	c := &EtcdEndpointsController{
+		operatorClient:  operatorClient,
+		nodeLister:      nodeInformer.Lister(),
+		configmapLister: configmapsInformer.Lister(),
+		configmapClient: kubeClient.CoreV1(),
+	}
+	return factory.New().ResyncEvery(time.Minute).WithInformers(
+		operatorClient.Informer(),
+		configmapsInformer.Informer(),
+		nodeInformer.Informer(),
+	).WithSync(c.sync).ToController("EtcdEndpointsController", eventRecorder.WithComponentSuffix("etcd-endpoints-controller"))
+}
+
+func (c *EtcdEndpointsController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	err := c.syncConfigMap(ctx, syncCtx.Recorder())
+
+	if err != nil {
+		_, _, updateErr := v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:    "EtcdEndpointsDegraded",
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "ErrorUpdatingEtcdEndpoints",
+			Message: err.Error(),
+		}))
+		if updateErr != nil {
+			syncCtx.Recorder().Warning("EtcdEndpointsErrorUpdatingStatus", updateErr.Error())
+		}
+		return err
+	}
+
+	_, _, updateErr := v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+		Type:   "EtcdEndpointsDegraded",
+		Status: operatorv1.ConditionFalse,
+		Reason: "EtcdEndpointsUpdated",
+	}))
+	if updateErr != nil {
+		syncCtx.Recorder().Warning("EtcdEndpointsErrorUpdatingStatus", updateErr.Error())
+		return updateErr
+	}
+	return nil
+}
+
+func (c *EtcdEndpointsController) syncConfigMap(ctx context.Context, recorder events.Recorder) error {
+	required := configMapAsset()
+
+	// create endpoint addresses for each node
+	nodes, err := c.nodeLister.List(labels.Set{"node-role.kubernetes.io/master": ""}.AsSelector())
+	if err != nil {
+		return fmt.Errorf("unable to list expected etcd member nodes: %v", err)
+	}
+	endpointAddresses := map[string]string{}
+	for _, node := range nodes {
+		var nodeInternalIP string
+		for _, nodeAddress := range node.Status.Addresses {
+			if nodeAddress.Type == corev1.NodeInternalIP {
+				nodeInternalIP = nodeAddress.Address
+				break
+			}
+		}
+		if len(nodeInternalIP) == 0 {
+			return fmt.Errorf("unable to determine internal ip address for node %s", node.Name)
+		}
+		endpointAddresses[nodeInternalIP] = ""
```

Comment:
@hexfusion make sure these are right for ipv6.  I expected slightly different code above.

Reply:
I'm now encoding the keys.